New module: xmlrpc_client

stemid opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
I have an API internally where this module helps a lot because it lets me use ansible for the entire deployment process. 

It's the most basic xmlrpc API client that forwards any call to the API along with positional arguments and keyword arguments. This way the ansible user can decide how to use the xmlrpc api along with the relevant API docs.

Looking at the webfaction modules for example I can see that a basic xmlrpc client module would be able to perform their tasks as well but with less abstractions between the user and the API.

I wasn't sure which category to put the module in. I'm sure people will have issue with placing it under commands, in that case just know that I have no problem with moving it to another category if a more suitable one can be found.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/modules/commands/xmlrpc_client
##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (xmlrpc_client-module 5224579d9e) last updated 2017/08/11 16:08:55 (GMT +200)
  config file = None
  configured module search path = ['/home/stemid/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/stemid/Utveckling/ansible/lib/ansible
  executable location = /home/stemid/Utveckling/ansible/bin/ansible
  python version = 3.6.2 (default, Jul 19 2017, 13:09:21) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
I have run sanity checks and there are no more issues with my code but some sanity checks in ansible-test still fail because of other older modules in the repo. So I expect ansibot to start posting those here soon. What can I do when all my own sanity checks are resolved?

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

stemid committed:
new module: xmlrpc_client for issuing xmlrpc commands to APIs.

stemid committed:
fixed the three last failing sanity checks for this file.

Qalthos committed:
Fix nxos_snmp_contact, others (#28000)

* Fix nxos_snmp_contact

Fixes #27714

* Fix nxos_snmp_community

Fixes #27711

* Fix nxos_install_os

Fixes #27604

* Update nxos_ntp_auth

Fixes #27521

* Fix nxos_igmp_snooping

Fixes #27501

schlueter committed:
Don't compare group_vars paths with bytestrings (#27922)

* Don't compare group_vars paths with bytestrings

* Compare spath with unicode string in VarsModule

stemid committed:
more uniform formatting of examples, 2 spaces indent fixed on all lines.

rcarrillocruz committed:
Replace username for name in eos_user integration tests (#28067)

username param is read by eos action plugion load_provider, if present
it is set as the play context remote user.
That means if you have a play remote_user: admin it won't be used, but instead
whatever is put in the task argument username.
Changing the tests to use name fixes the issue, we may want to remove username
as valid module argument in other change.

lmprice committed:
Refactor E-Series Hosts module to use module_utils (#20893)

Refactor the NetApp E-Series module to utlize the common module_utils
and doc_fragments.

trstringer committed:
Add support for Azure 2.0.0 (#27920)

* Adapt azure_rm_resource_group to azure 2.0.0 + azure Cli support

* Fix exceptions in Azure ARM plugins

* update azure_rm_networkinterface documention to reflect required params

* change state param to not required for docs in azure_rm_subnet

* fix import to reflect azure==2.0.0 changes

* add aliases and fix docs for azure_rm_storageblob

* add resource_group_name alias to azure_rm_storageaccount_facts

* fix import bug due to change in azure==2.0.0

* fix args bug and enum modules issue

* update docs to reflect azure==2.0.0

* pin management clients to a specific api_version

* update docs to reflect the new azure-ansible-base python package

* add fallback for older api resource group listing

* rework azure dependencies installation

* refactor path joining to a cross-plat solution

s-hertel committed:
[cloud] Make ec2_asg more resilient using AWSRetry around boto3 calls (#27598)

* Add AWSRetry to ec2_asg

* Paginate describing ASGs and launch configurations

pass connection to delete_asg

Fix a couple little bugs

* Use boto3's pagination build_full_result()

ogenstad committed:
Fixes for broken asa_config module (#27218)

* Fixes for broken asa_config module

* strip() prompt before checking

alvaroaleman committed:
Allow the Openstack dynamic inventory to discover ansible_user (#17944)

privateip committed:
fixes command to be sent over nxapi for nxos_file_copy (#28029)

fixes #27347

trishnaguha committed:
fix nxos_snapshot (#27993)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

jmcgill298 committed:
fix issue with trailing white space when logging into asa (#28073)

abadger committed:
Fix for broken azure requirements

Older setuptools does not understand ~= versioning.  Guess at a range
that could be compatible to unbreak CI.

mattclay committed:
Update ansible-test handling of setup.py changes.

bcoca committed:
validate include_role args (#28077)

* validate include_role args

* nicer error msg

* removed unused static

* updated as per feedback

* pepe ATE

s-hertel committed:
[cloud] Allow S3 module to correctly delete objects by groups of <= 1000 (#28068)

chriskarel committed:
Fix idempotency for Unix permissions in zip files. (#24580)

* Fix idempotency for Unix permissions in zip files.

This fix prevents the unarchive module from reporting 'changed' when a zipfile contains items with Unix permissions that differ from the system default.

* Update zip unarchive tests.

Additional tests for the unarchive module with zip files:
- Test file in zip archive with non-default permissions
- Test file added to zip archive with Windows permissions

* Additional fix for mixed win/unix archives.

  Turns out my original fix fails under some mixed archives, as setting the umask to zero can be applied to those files.  This creates a per-file umask variable, so a mix of permission types don't cause problems.

* CI Checks

CI checks for archives with:
* non default Unix permissions
* Windows permissions


* Workaround for BSD differences.

Using Zipinfo due to lack of support in BSD unzip.
Permissions handling is also different in BSD -- always applies UMASK to file permissions.

* Added checks for creating directories and SSH keys for existing users.

stemid committed:
Typo in module example code (#28064)

In the DOCUMENTATION constant there is a typo on line 85 of this rst file.

dagwieers commented on lib/ansible/modules/commands/xmlrpc_client.py:
```diff
@@ -0,0 +1,167 @@
+    args:
+        description:
+            - A list of arguments for the RPC call. Placed first in any RPC call.
+        required: false
```
This is not needed for parameters that are not required. Guideline dictates to leave it out.

dagwieers commented on lib/ansible/modules/commands/xmlrpc_client.py:
```diff
@@ -0,0 +1,167 @@
+    kwargs:
+        description:
+            - A dictionary of arguments for the RPC call. Comes after the args list.
+        required: false
```
This is not needed for parameters that are not required. Guideline dictates to leave it out.

dagwieers commented on lib/ansible/modules/commands/xmlrpc_client.py:
```diff
@@ -0,0 +1,167 @@
+    module_args = dict(
+        url=dict(type='str', required=True),
+        path=dict(type='str', required=True),
+        args=dict(type='list', required=False, default=[]),
```
Here also `required=False` is not needed for parameters that are not required. Guideline dictates to leave it out.

dagwieers commented on lib/ansible/modules/commands/xmlrpc_client.py:
```diff
@@ -0,0 +1,167 @@
+        url=dict(type='str', required=True),
+        path=dict(type='str', required=True),
+        args=dict(type='list', required=False, default=[]),
+        kwargs=dict(type='dict', required=False, default={})
```
Here also `required=False` is not needed for parameters that are not required. Guideline dictates to leave it out.

dagwieers reviewed (changes requested):
Looking good.

lingfish committed:
Documentation fixes. (#28047)

* Documentation fixes.

* Update win_route.py

dagwieers committed:
Add new maintainer for ACI

dagwieers committed:
Fix indentation

Don't use the Github editor :-/

jmcgill298 committed:
update aci module util with methods to support check_mode and idempotency (#28089)

jmcgill298 committed:
Update aci_tenant to support check_mode (#28090)

mattclay committed:
Migrate roles from ansible/azure-testing (#28074)

* Migrate roles from ansible/azure-testing as-is.
* Fix yamllint issues.
* Remove unused binary file.

abadger committed:
Fix symbolic mode usage in unarchive with zip files

Fixes #21090

mattclay committed:
Add Azure support to ansible-test. (#28092)

* Initial Azure cloud testing support.
* Add missing Azure requirements.
* Fix test handling of setup and requirements.
* Update Azure cloud plugin.
* Add setup_azure role for integration tests.
* Update minimal Azure integration test sample.

abadger committed:
Optimize template (#28044)

* Optimize template

* In fixing template to handle diff correctly #24477, I introduced more
  round trips to the remote end which slowed things down  The new code now
  uses one fewer round trips than the old code.
* Reimplement a large part of template by calling the copy action plugin
  instead of doing it in template's code.  This reduces the code in
  template and gives us one place to fix bugs and optimize.
* Add a follow parameter to template that mirrors the follow parameters
  for file and copy.
* Fix copy's diff handling (probably broken in my rewrite for in 2.4
  development)
* Adjusted when copy creates tmp dirs to rduce round trips in copy and
  template.

Fixes #27956

bcoca committed:
telnet action plugin (#27983)

* draft telnet action plugin

docs in module

* updated per feedback

* fixed imports

* pep8 fixes

* added meta import even though its useless

abadger committed:
Remove wildcard imports and get_exception calls

Fixed module_utils

privateip committed:
update telnet module (#28100)

* accepts list of prompts to expect
* updates doc strings

Cerebus committed:
ini_file: Match tabs before options (#28102)

mattclay committed:
Clean up and migrate Azure tests. (#28103)

* Remove placeholder Azure test.
* Migrate Azure tests to ansible-test.
* Initial cleanup on remaining legacy Azure tests.

ar7z1 committed:
rabbitmq_plugin for Windows (#28017)

* Fixes #28016: rabbitmq_plugin for Windows

* Disable documentation check for rabbitmq_plugin.ps1

* Renamed rabbitmq_plugin -> win_rabbitmq_plugin

* Fixed the documentation after review

* Fixed 'RETURN' section

* Fixed docs for original module

* Added dots to original module docs.

s-hertel committed:
iam certificate facts (duplicate): add iam_cert_facts as alias for iam_server_certificate_facts (#25387)

* add iam_cert_facts as alias for iam_server_certificate_facts
* remove from legacy files

stemid committed:
fixing requested code changes, added clarifying comment for tryck block.

stemid commented:
@dagwieers Requested changes made, committed and sanity checks still report ok for this module.

stemid commented:
I made a mistake with the rebase and now we have 20 unnecessary participants so I'm closing this and opening a new one. Sorry for the confusion.

dagwieers commented:
@stemid You can always fix that by force-pushing the (corrected) branch. The pull-request will be adapted. (Often you can correct your branch using `git rebase -i HEAD~40` removing all commits unrelated to your change up to the point the branch was still in sync with devel.)
